chart/tests/test_redis.py

```diff
@@ -0,0 +1,205 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import re
+import unittest
+from base64 import b64decode
+from subprocess import CalledProcessError
+from typing import Optional, Pattern, Set, Tuple
+
+from parameterized import parameterized
+
+from tests.helm_template_generator import render_chart
+
+
+RELEASE_NAME_REDIS = "TEST-REDIS"
+
+REDIS_OBJECTS = {
+    "NETWORK_POLICY": ("NetworkPolicy", f"{RELEASE_NAME_REDIS}-redis-policy"),
+    "SERVICE": ("Service", f"{RELEASE_NAME_REDIS}-redis"),
+    "STATEFUL_SET": ("StatefulSet", f"{RELEASE_NAME_REDIS}-redis"),
+    "SECRET_PASSWORD": ("Secret", f"{RELEASE_NAME_REDIS}-redis-password"),
+    "SECRET_BROKER_URL": ("Secret", f"{RELEASE_NAME_REDIS}-broker-url"),
+}
+
+CELERY_EXECUTORS_PARAMS = [("CeleryExecutor",), ("CeleryKubernetesExecutor",)]
+
+
+class RedisTest(unittest.TestCase):
+    def check_redis_objects_created_by_chart(
```

Comment:
This complicates the tests a bit. What do you think to break this function into several functions so that each function is responsible for one assertion?

Reply:
I wasn't sure about that part when coding it. Could you be a bit more specific regarding what I should aim for ?

* Simply split this helper into multiple more specific helpers,
* Or redisign a bit the tests so that each "test" perform by pytest corresponds to one of the main parts of this big helper
* Or something else ?

Reply:
Basically, I would like to simplify this text so that it does not contain as many loops and that the main logic of the test is contained in the test function, not in the helper functions.

Here is an example that creates a simple assertion based on comparing two sets instead of using a list and negation.
```
From 12fff535b966e1b1fc881cadecd30f09a70a1df0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kamil=20Bregu=C5=82a?= <kamil.bregula@polidea.com>
Date: Sun, 1 Nov 2020 12:53:57 +0100
Subject: Simplify assertion

---
 chart/tests/test_redis.py | 68 +++++++++++++++++++++++++++------------
 1 file changed, 47 insertions(+), 21 deletions(-)

diff --git a/chart/tests/test_redis.py b/chart/tests/test_redis.py
index 117487127..6e365437f 100644
--- a/chart/tests/test_redis.py
+++ b/chart/tests/test_redis.py
@@ -18,7 +18,7 @@ import re
 import unittest
 from base64 import b64decode
 from subprocess import CalledProcessError
-from typing import Optional, Pattern, Set, Tuple
+from typing import Optional, Pattern
 
 from parameterized import parameterized
 
@@ -42,7 +42,6 @@ class RedisTest(unittest.TestCase):
     def check_redis_objects_created_by_chart(
         self,
         k8s_objects,
-        expected_obj_keys_not_present: Set[Tuple[str, str]],
         expected_redis_password_regex: Optional[Pattern] = None,
         expected_redis_broker_url_regex: Optional[Pattern] = None,
         expected_broker_url_secret_name: str = REDIS_OBJECTS["SECRET_BROKER_URL"][1],
@@ -50,17 +49,7 @@ class RedisTest(unittest.TestCase):
         """
         Helper to perform checks for the redis integration on the rendered objects
         """
-        k8s_obj_by_key = {
-            (k8s_object["kind"], k8s_object["metadata"]["name"]): k8s_object for k8s_object in k8s_objects
-        }
-
-        # check that the expected objects are created and that no extra objects are created
-        expected_obj_key_present = set(REDIS_OBJECTS.values()) - expected_obj_keys_not_present
-        for obj_key in expected_obj_key_present:
-            self.assertIn(obj_key, k8s_obj_by_key.keys())
-
-        for obj_key in expected_obj_keys_not_present:
-            self.assertNotIn(obj_key, k8s_obj_by_key)
+        k8s_obj_by_key = self.prepare_k8s_lookup_dict(k8s_objects)
 
         # check the password value in the secret
         if expected_redis_password_regex is not None:
@@ -95,6 +84,12 @@ class RedisTest(unittest.TestCase):
             "Deployment", "scheduler"
         ), "Couldn't verify that AIRFLOW__CELERY__BROKER_URL is properly set in scheduler."
 
+    def prepare_k8s_lookup_dict(self, k8s_objects):
+        k8s_obj_by_key = {
+            (k8s_object["kind"], k8s_object["metadata"]["name"]): k8s_object for k8s_object in k8s_objects
+        }
+        return k8s_obj_by_key
+
     @parameterized.expand(CELERY_EXECUTORS_PARAMS)
     def test_redis_by_chart_default(self, executor):
         k8s_objects = render_chart(
@@ -107,7 +102,6 @@ class RedisTest(unittest.TestCase):
         )
         self.check_redis_objects_created_by_chart(
             k8s_objects,
-            expected_obj_keys_not_present=set(),
             expected_redis_password_regex=re.compile(r"\w+"),
             expected_redis_broker_url_regex=re.compile(fr"redis://:\w+@{RELEASE_NAME_REDIS}-redis:6379/0"),
         )
@@ -122,9 +116,22 @@ class RedisTest(unittest.TestCase):
                 "redis": {"enabled": True, "password": "test-redis-password"},
             },
         )
+        k8s_obj_by_key = self.prepare_k8s_lookup_dict(k8s_objects)
+
+        # assert resources created
+        self.assertEqual(
+            set(REDIS_OBJECTS.values()).intersection(set(k8s_obj_by_key.keys())),
+            {
+                ('Secret', 'TEST-REDIS-redis-password'),
+                ('NetworkPolicy', 'TEST-REDIS-redis-policy'),
+                ('StatefulSet', 'TEST-REDIS-redis'),
+                ('Service', 'TEST-REDIS-redis'),
+                ('Secret', 'TEST-REDIS-broker-url')
+            },
+        )
+
         self.check_redis_objects_created_by_chart(
             k8s_objects,
-            expected_obj_keys_not_present=set(),
             expected_redis_password_regex=re.compile(r"test-redis-password"),
             expected_redis_broker_url_regex=re.compile(
                 fr"redis://:test-redis-password@{RELEASE_NAME_REDIS}-redis:6379/0"
@@ -159,12 +166,20 @@ class RedisTest(unittest.TestCase):
                 },
             },
         )
+        k8s_obj_by_key = self.prepare_k8s_lookup_dict(k8s_objects)
+
+        # assert resources created
+        self.assertEqual(
+            set(REDIS_OBJECTS.values()).intersection(set(k8s_obj_by_key.keys())),
+            {
+                ('NetworkPolicy', 'TEST-REDIS-redis-policy'),
+                ('StatefulSet', 'TEST-REDIS-redis'),
+                ('Service', 'TEST-REDIS-redis'),
+            },
+        )
+
         self.check_redis_objects_created_by_chart(
             k8s_objects,
-            expected_obj_keys_not_present={
-                REDIS_OBJECTS["SECRET_PASSWORD"],
-                REDIS_OBJECTS["SECRET_BROKER_URL"],
-            },
             expected_broker_url_secret_name="test-redis-broker-url-secret-name",
         )
 
@@ -181,9 +196,16 @@ class RedisTest(unittest.TestCase):
                 "redis": {"enabled": False},
             },
         )
+        k8s_obj_by_key = self.prepare_k8s_lookup_dict(k8s_objects)
+
+        # assert resources created
+        self.assertEqual(
+            set(REDIS_OBJECTS.values()).intersection(set(k8s_obj_by_key.keys())),
+            {('Secret', 'TEST-REDIS-broker-url')},
+        )
+
         self.check_redis_objects_created_by_chart(
             k8s_objects,
-            expected_obj_keys_not_present=set(REDIS_OBJECTS.values()) - {REDIS_OBJECTS["SECRET_BROKER_URL"]},
             expected_redis_broker_url_regex=re.compile("redis://redis-user:password@redis-host:6379/0"),
         )
 
@@ -198,8 +220,12 @@ class RedisTest(unittest.TestCase):
                 "redis": {"enabled": False},
             },
         )
+        k8s_obj_by_key = self.prepare_k8s_lookup_dict(k8s_objects)
+
+        # assert resources created
+        self.assertEqual(set(REDIS_OBJECTS.values()).intersection(set(k8s_obj_by_key.keys())), set())
+
         self.check_redis_objects_created_by_chart(
             k8s_objects,
-            expected_obj_keys_not_present=set(REDIS_OBJECTS.values()),
             expected_broker_url_secret_name="redis-broker-url-secret-name",
         )
-- 
2.28.0

```

Reply:
I would be happy if you would make separate functions that perform assertions for environment variables and secrets.
```
        self.assert_secrets(
            k8s_objects,
            expected_redis_password_regex=r"\w+",
            expected_redis_broker_url_regex=fr"redis://:\w+@{RELEASE_NAME_REDIS}-redis:6379/0",
        )
        self._assert_broker_url_envs(k8s_obj_by_key, 'redis-broker-url-secret-name')
```

Reply:
I've done two fixups on the tests to address your comments ; I hope it's be better now.